### PR TITLE
Update FileInfo when the file is committed

### PR DESCRIPTION
**Pull request by jiacheliu3**
Fixes https://github.com/Alluxio/alluxio/issues/13128
When this guard is added the issue does not happen anymore.

The trade-off is, when we `getFileInfo` on a file that is currently being written, the number of `BlockId` and `FileBlockInfo` we get will mismatch. I don't think that is a problem.

### CODE REVIEW

**Comment by jiacheliu3** [comment]
I should probably add an integration test

**Comment by jiacheliu3** [comment]
@apc999 One potential problem is, we might have been relying on this rehydration to commit files that are not properly closed. When a `FileOutStream` is not closed (and the process just dies), on the next `getStatus` the file is implicitly committed and we don't experience an issue (we are not aware this happened at all). 

Now with this change, a `FileOutStream` that is not closed will have the file in the completed status forever? Is that a problem? 

In my option, we should not rely on the rehydration logic here to handle incomplete files anyway. So I'm in favor of this change.

File: tests/src/test/java/alluxio/client/fs/io/FileOutStreamIntegrationTest.java [hunk 1]
```diff
@@ -247,4 +248,31 @@ public void cancelWrite() throws Exception {
...
+    AlluxioURI path = new AlluxioURI(PathUtils.uniqPath());
+    try (FileOutStream os = mFileSystem.createFile(path, CreateFilePOptions.newBuilder()
+            .setWriteType(mWriteType.toProto()).setRecursive(true).build())) {
+      for (int i = 0; i < 3; i++) {
+        os.write(BufferUtils.getIncreasingByteArray(i * BLOCK_SIZE_BYTES, BLOCK_SIZE_BYTES));
+        // Fetch file status when the stream is still open
+        URIStatus status = mFileSystem.getStatus(path);
+        if (!mWriteType.isThrough()) {
+          // When the writeType is THROUGH, we only see the blocks when the file is committed
+          // When the file is not committed, we do not see the FileBlockInfo
```
Comment (jiacheliu3) [inline tests/src/test/java/alluxio/client/fs/io/FileOutStreamIntegrationTest.java:263]: Before this change, we see `FileBlockInfo` but that does not necessarily match the number of block IDs. And all `FileBlockInfo` will have 0 length.
After this change, we see 0 `FileBlockInfo` before the file is committed.

File: tests/src/test/java/alluxio/client/fs/io/FileOutStreamIntegrationTest.java [hunk 2]
```diff
@@ -247,4 +248,31 @@ public void cancelWrite() throws Exception {
       Assert.assertEquals(0, worker.getUsedBytes());
     }
   }
+
+  @Test
+  public void getStatusBeforeClose() throws Exception {
```
Comment (jiacheliu3) [inline tests/src/test/java/alluxio/client/fs/io/FileOutStreamIntegrationTest.java:253]: Before this change, this integration test does not pass for `CACHE_THROUGH`. The stream will fail at `close()`.

**Review by apc999 — APPROVED** [review]

**Review by calvinjia — APPROVED** [review]

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by jiacheliu3** [comment]
Failed test is irrelevant to this PR
```
04:06:35.995 [INFO] Running alluxio.server.ft.journal.MultiMasterJournalTest
Error: 7.137 [ERROR] Aborted (core dumped)
04:07:07.796 [INFO] Running alluxio.server.ft.journal.MultiProcessCheckpointTest
04:08:06.265 [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 58.464 s - in alluxio.server.ft.journal.MultiProcessCheckpointTest
04:08:08.498 [INFO] Running alluxio.server.ft.journal.TableMasterJournalIntegrationTest
Error: 0.192 [ERROR] WARNING: An illegal reflective access operation has occurred
Error: 0.202 [ERROR] WARNING: Illegal reflective access by org.powermock.reflect.internal.WhiteboxImpl (file:/home/jenkins/.m2/repository/org/powermock/powermock-reflect/2.0.7/powermock-reflect-2.0.7.jar) to method java.lang.Object.clone()
Error: 0.209 [ERROR] WARNING: Please consider reporting this to the maintainers of org.powermock.reflect.internal.WhiteboxImpl
Error: 0.210 [ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Error: 0.220 [ERROR] WARNING: All illegal access operations will be denied in a future release
```

**Comment by jiacheliu3** [comment]
To be cherry-picked to 2.4 and 2.5 after the test issue is resolved.

**Comment by jiacheliu3** [comment]
The test error mysteriously disappeared on the 6th or 7th run.

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-2.4 please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-2.5 please

**Pull request merged**